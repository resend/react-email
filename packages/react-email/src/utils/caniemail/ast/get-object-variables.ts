import type { Node } from '@babel/traverse';
import traverse from '@babel/traverse';
import type { AST } from '../../../actions/get-insights-for-email';

export interface Position {
  line: number;
  column: number;
  index: number;
}

export const convertLocationIntoObject = (
  location: SourceLocation,
): SourceLocation => {
  return {
    start: {
      line: location.start.line,
      column: location.start.column,
      index: location.start.index,
    },
    end: {
      line: location.end.line,
      column: location.end.column,
      index: location.end.index,
    },
    filename: location.filename,
    identifierName: location.identifierName,
  };
};

export interface SourceLocation {
  start: Position;
  end: Position;
  filename: string;
  identifierName: string | undefined | null;
}

type ObjectProperty = Node & { type: 'ObjectProperty' };

export type ObjectVariables = Record<string, ObjectProperty[]>;

export const getObjectVariables = (ast: AST) => {
  const objectVariables: ObjectVariables = {};
  traverse(ast, {
    ObjectExpression(nodePath) {
      if (nodePath.parent.type === 'VariableDeclarator') {
        if (nodePath.parent.id.type === 'Identifier') {
          const variableName = nodePath.parent.id.name;
          const properties: ObjectProperty[] = [];
          for (const property of nodePath.node.properties) {
            if (property.type === 'ObjectProperty') {
              properties.push(property);
            }
          }
          objectVariables[variableName] = properties;
        }
      }
    },
  });

  return objectVariables;
};
