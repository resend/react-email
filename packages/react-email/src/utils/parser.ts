import * as acorn from 'acorn';
import typescript from 'acorn-typescript';

export const Parser = acorn.Parser.extend(
  // @ts-expect-error
  typescript({
    allowSatisfies: true,
  }),
);

export type AST = ReturnType<typeof Parser.parse>;
