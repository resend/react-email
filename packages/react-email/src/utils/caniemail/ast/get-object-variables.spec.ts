import { Parser } from '../../parser';
import { getObjectVariables } from './get-object-variables';

test('getObjectVariables()', () => {
  const reactCode = `
<Button style={buttonStyle}>Click me</Button>

const buttonStyle = {
  borderRadius: '5px',
};
`;
  const ast = Parser.parse(reactCode, {
    ecmaVersion: 'latest',
    sourceType: 'module',
  });
  expect(getObjectVariables(ast)).toMatchSnapshot();
});
