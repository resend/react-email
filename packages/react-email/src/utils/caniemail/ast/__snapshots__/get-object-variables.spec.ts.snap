// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getObjectVariables() 1`] = `
{
  "buttonStyle": [
    Node {
      "computed": false,
      "end": 91,
      "key": Node {
        "end": 84,
        "loc": SourceLocation {
          "end": Position {
            "column": 14,
            "index": 84,
            "line": 5,
          },
          "filename": undefined,
          "identifierName": "borderRadius",
          "start": Position {
            "column": 2,
            "index": 72,
            "line": 5,
          },
        },
        "name": "borderRadius",
        "start": 72,
        "type": "Identifier",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 21,
          "index": 91,
          "line": 5,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 2,
          "index": 72,
          "line": 5,
        },
      },
      "method": false,
      "shorthand": false,
      "start": 72,
      "type": "ObjectProperty",
      "value": Node {
        "end": 91,
        "extra": {
          "raw": "'5px'",
          "rawValue": "5px",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 21,
            "index": 91,
            "line": 5,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 16,
            "index": 86,
            "line": 5,
          },
        },
        "start": 86,
        "type": "StringLiteral",
        "value": "5px",
      },
    },
  ],
}
`;
